### PR TITLE
 playwright: fix firefox 

### PR DESCRIPTION
**Pull request by DGollings**
###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by DGollings** [comment]
Opened this PR as requested by @SuperSandro2000, this fixes the currently broken firefox binary. But be warned, I have a very basic understanding of Nix, feedback welcome

**Comment by xanderio** [comment]
Your commit message doesn't fit with [Contributing.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md). In this case I think it should be `playwright: fix firefox`

**Comment by xanderio** [comment]
@ofborg build playwright

**Comment by DGollings** [comment]
done @xanderio

**Comment by xanderio** [comment]
Result of `nixpkgs-review pr 197899` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by xanderio** [comment]
Interesting `nixpkgs-review` doesn't see any package rebuilds, not sure what is going on.

File: pkgs/development/python-modules/playwright/default.nix [hunk 1]
```diff
@@ -79,6 +83,22 @@ let
...
   };
 
+    browser_revs =
+    let
+      file = fetchurl {
+        url =
+          "https://raw.githubusercontent.com/microsoft/playwright/v${driverVersion}/packages/playwright-core/browsers.json";
+        sha256 = "11a1n65l43nfyjn0qrsjjyjl7psvqa67k4kiv8x624faga4zl3mk";
+      };
+      raw_data = fromJSON (readFile file);
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/playwright/default.nix:93]:
```suggestion
      raw_data = lib.importJSON file;
```

File: pkgs/development/python-modules/playwright/default.nix [hunk 2]
```diff
@@ -21,7 +22,10 @@
 }:
 
 let
+  inherit (builtins) fromJSON readFile listToAttrs;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/playwright/default.nix:25]:
```suggestion
```

File: pkgs/development/python-modules/playwright/default.nix [hunk 3]
```diff
@@ -21,7 +22,10 @@
...
   inherit (stdenv.hostPlatform) system;
+  selectSystem = attrs:
+  attrs.${system} or (throw "Unsupported system: ${system}");
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/playwright/default.nix:28]:
```suggestion
  selectSystem = attrs: attrs.${system} or (throw "Unsupported system: ${system}");
```

File: pkgs/development/python-modules/playwright/default.nix [hunk 4]
```diff
@@ -79,6 +83,22 @@ let
     };
   };
 
+    browser_revs =
+    let
+      file = fetchurl {
+        url =
+          "https://raw.githubusercontent.com/microsoft/playwright/v${driverVersion}/packages/playwright-core/browsers.json";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/playwright/default.nix:90]:
```suggestion
        url = "https://raw.githubusercontent.com/microsoft/playwright/v${driverVersion}/packages/playwright-core/browsers.json";
```

File: pkgs/development/python-modules/playwright/default.nix [hunk 5]
```diff
@@ -79,6 +83,22 @@ let
...
+        sha256 = "11a1n65l43nfyjn0qrsjjyjl7psvqa67k4kiv8x624faga4zl3mk";
+      };
+      raw_data = fromJSON (readFile file);
+    in
+    listToAttrs (map
+      ({ name, revision, ... }: {
+        inherit name;
+        value = revision;
+      })
+      raw_data.browsers);
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/playwright/default.nix:100]:
```suggestion
    listToAttrs (map (
      { name, revision, ... }: {
        inherit name;
        value = revision;
      }
    ) raw_data.browsers);
```

File: pkgs/development/python-modules/playwright/default.nix [hunk 6]
```diff
@@ -105,6 +125,20 @@ let
     fontconfig = makeFontsConf {
       fontDirectories = [];
     };
+  suffix = selectSystem {
+    # Not sure how other system compatibility is, needs trial & error
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/playwright/default.nix:129]:
```suggestion
```
thats fine :)

File: pkgs/development/python-modules/playwright/default.nix [hunk 7]
```diff
@@ -105,6 +125,20 @@ let
...
+    x86_64-linux = "ubuntu-20.04";
+    aarch64-linux = "ubuntu-20.04-arm64";
+    x86_64-darwin = "mac";
+    aarch64-darwin = "mac-arm64";
+  };
+
+  upstream_firefox = fetchzip {
+    url =
+      "https://playwright.azureedge.net/builds/firefox/${browser_revs.firefox}/firefox-${suffix}.zip";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/playwright/default.nix:138]:
```suggestion
    url = "https://playwright.azureedge.net/builds/firefox/${browser_revs.firefox}/firefox-${suffix}.zip";
```

File: pkgs/development/python-modules/playwright/default.nix [hunk 8]
```diff
@@ -125,9 +159,29 @@ let
       --set SSL_CERT_FILE /etc/ssl/certs/ca-bundle.crt \
       --set FONTCONFIG_FILE ${fontconfig}
   '' + lib.optionalString withFirefox ''
-    FIREFOX_REVISION=$(jq -r '.browsers[] | select(.name == "firefox").revision' $BROWSERS_JSON)
-    mkdir -p $out/firefox-$FIREFOX_REVISION/firefox
-    ln -s ${firefox}/bin/firefox $out/firefox-$FIREFOX_REVISION/firefox/firefox
+    firefoxoutdir=$out/firefox-${browser_revs.firefox}/firefox
+    mkdir -p $firefoxoutdir
+    cp -r ${upstream_firefox}/* $firefoxoutdir/
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/playwright/default.nix:164]: We should think about symlinks or xorg.lndir

File: pkgs/development/python-modules/playwright/default.nix [hunk 9]
```diff
@@ -125,9 +159,29 @@ let
...
+
+    # patchelf the binary
+    wrapper="${firefox-bin}/bin/firefox"
+    binary="$(readlink -f $(<"$wrapper" grep '^exec ' | grep -o -P '/nix/store/[^"]+' | head -n 1))"
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/playwright/default.nix:168]: This is at a fixed location? Or can we get that from the firefox attr via passthru?

File: pkgs/development/python-modules/playwright/default.nix [hunk 10]
```diff
@@ -125,9 +159,29 @@ let
...
+
+    interpreter="$(patchelf --print-interpreter "$binary")"
+    rpath="$(patchelf --print-rpath "$binary")"
+
+    find $firefoxoutdir/ -executable -type f | while read i; do
+      chmod u+w "$i"
+      [[ $i == *.so ]] || patchelf --set-interpreter "$interpreter" "$i"
+      patchelf --set-rpath "$rpath" "$i"
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/playwright/default.nix:176]:
```suggestion
    find $firefoxoutdir/ -executable -type f | while read file; do
      chmod u+w "$file"
      [[ $file == *.so ]] || patchelf --set-interpreter "$(patchelf --print-interpreter "$binary")" "$file"
      patchelf --set-rpath "$(patchelf --print-rpath "$binary")" "$file"
```

File: pkgs/development/python-modules/playwright/default.nix [hunk 11]
```diff
@@ -125,9 +159,29 @@ let
...
+      chmod u-w "$i"
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/playwright/default.nix:177]:
```suggestion
```
fixed by nix

File: pkgs/development/python-modules/playwright/default.nix [hunk 12]
```diff
@@ -125,9 +159,29 @@ let
...
+    done
+
+    # create the wrapper script
+    rm $firefoxoutdir/firefox
+    <"$wrapper" grep -vE '^exec ' > $firefoxoutdir/firefox
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/playwright/default.nix:182]: I am not sure what this line does

File: pkgs/development/python-modules/playwright/default.nix [hunk 13]
```diff
@@ -125,9 +159,29 @@ let
...
+    echo "exec \"$firefoxoutdir/firefox-bin\" \"\$@\"" >> $firefoxoutdir/firefox
+    chmod a+x $firefoxoutdir/firefox
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/playwright/default.nix:184]: We can't use wrapProgram?

**Comment by teto** [comment]
would you mind updating the PR following up on https://github.com/NixOS/nixpkgs/pull/223382 ? I am willing to help review/get merged

**Comment by DGollings** [comment]
@teto , I'll have a look when I have time, but am quite busy at the moment. Hopefully I'll be able to re-add Firefox support soon

**Comment by teto** [comment]
firefox support was added in https://github.com/NixOS/nixpkgs/pull/298944 and this is old. Thanks for the effort. Closing

**Pull request closed without merging**